servant-minimalist.hsfiles

```diff
@@ -0,0 +1,232 @@
+  - ChangeLog.md
+
+# Metadata used when publishing your package
+# synopsis:            Short description of your package
+# category:            {{category}}{{^category}}Web{{/category}}
+
+# To avoid duplicated efforts in documentation and dealing with the
+# complications of embedding Haddock markup inside cabal files, it is
+# common to point users to the README.md file.
+description:         Please see the README on GitHub at <https://github.com/{{github-username}}{{^github-username}}githubuser{{/github-username}}/{{name}}#readme>
+
+dependencies:
+  - base >= 4.7 && < 5
+  - aeson
+  - servant-server
+  - wai
+  - warp
+
+library:
+  source-dirs: src
+
+executables:
+  {{name}}-exe:
+    main:                Main.hs
+    source-dirs:         app
+    ghc-options:
+      - -threaded
+      - -rtsopts
+      - -with-rtsopts=-N
+    dependencies:
+      - {{name}}
+
+tests:
+  {{name}}-test:
+    main:                Spec.hs
+    source-dirs:         test
+    ghc-options:
+      - -threaded
+      - -rtsopts
+      - -with-rtsopts=-N
+    dependencies:
+      - {{name}}
+      # Property based testing
+      - QuickCheck
+      - quickcheck-instances
+      # Unit testing
+      - HUnit
+      # Test framework (aggregation and discovery)
+      - tasty
+      - tasty-quickcheck
+      - tasty-hunit
+      - tasty-discover
+
+{-# START_FILE Setup.hs #-}
+import           Distribution.Simple
+
+main = defaultMain
+
+{-# START_FILE test/Spec.hs #-}
+{-# OPTIONS_GHC -F -pgmF tasty-discover -optF --tree-display
+ #-}
+-- Driver module responsible for discovering all other modules, and tests inside modules, using tasty-discover preprocessor.
+
+{-# START_FILE test/LibSpec.hs #-}
+module LibSpec where
+
+import           Data.Aeson
+import           Lib
+import           Test.QuickCheck
+import           Test.QuickCheck.Instances ()
+import           Test.Tasty.HUnit
+import           Test.Tasty.QuickCheck
+
+-- Property test
+test_findById_no_user =
+  testProperty "findById for non existing users" $ \(User uid _ _) -> uid /= 1 && uid /= 2 ==> findById uid === Nothing
+
+-- Unit test
+test_findById_users =
+  testCase "findById for existing users" $ do
+    findById (userId albert) @?= Just albert
+    findById (userId isaac) @?= Just isaac
+
+instance Arbitrary User where
+  arbitrary = User <$> arbitrary <*> arbitrary <*> arbitrary
+
+{-# START_FILE src/Lib.hs #-}
+{-# LANGUAGE DataKinds       #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeOperators   #-}
+
+module Lib
+  ( startApp
+  , app
+  , findById
+  , isaac
+  , albert
+  , User(..)
+  ) where
+
+import           Data.Aeson
+import           Data.Aeson.TH
+import           Data.List
+import           Data.Maybe
+import           Network.Wai
+import           Network.Wai.Handler.Warp
+import           Servant
+
+data User = User
+  { userId        :: Int
+  , userFirstName :: String
+  , userLastName  :: String
+  } deriving (Eq, Show)
+
+$(deriveJSON defaultOptions ''User)
```

Comment:
why TH, and not Generics?

Reply:
I don't have any strong opinion on this question. I usually use TH because it's less code to write/maintain in the end by developers (we often have to change only the Aeson Options, and we have to modify it twice with Generics, while only once with TH). 

To clarify, by Generics, I assume you mean
```haskell
deriving (Eq, Show, Generics)

instance ToJSON User

instance FromJSON User
```
Am I correct? Either way, I'm ok with both, just let me know which you prefer (or refuse).

Reply:
that can be written as

```haskell
  deriving (Eq, Show, Generic, ToJSON, FromJSON)
```

or more explicitly with [`DerivingStrategies`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-strategies)

```haskell
  deriving stock (Eq, Show, Generic)
  deriving anyclass (ToJSON, FromJSON)
```